### PR TITLE
Reorganize the NonlinearSystem hierarchy

### PR DESCRIPTION
**Pull request by fdkong**
### Complete each of the following items before creating a PR

Reorganize the NonlinearSystem hierarchy to allow us to derive EigenSystem from Libmesh. To achieve this goal, the following steps will be taken:

(1) Rename  NonlinearSystem to NonlinearSystemBase

(2) Create a new class NonlinearSystem deriving from the class NonlinearSystemBase

(3) Change the inheritance of the class MooseEigenSystem. MooseEigenSystem should  derive  from the class NonlinearSystemBase instead of the class NonlinearSystem

(4) Move some functions specified for the NonlinearSystem only from the NonlinearSystemBase  to the class NonlinearSystem

(5) Enable SystemTempl<EigenSystem> to bring SLEPc from Libmesh

In this PR, we finished steps (1) and (2).

Refs  #7398

### CODE REVIEW

File: framework/include/base/FEProblem.h [hunk 1]
```diff
@@ -154,6 +154,8 @@ class FEProblem :
 
   bool areCoupled(unsigned int ivar, unsigned int jvar);
 
+  bool hasConstraints() { return _has_constraints; }
```
Comment (YaqiWang) [inline framework/include/base/FEProblem.h]: make this const.

File: framework/include/base/NonlinearSystem.h [hunk 2]
```diff
@@ -12,618 +12,25 @@
...
-  bool _has_nodalbc_save_in;
-
-  /// If there is a nodal BC having diag_save_in
-  bool _has_nodalbc_diag_save_in;
-
-  void getNodeDofs(unsigned int node_id, std::vector<dof_id_type> & dofs);
 };
 
-#endif /* NONLINEARSYSTEM_H */
+#endif
```
Comment (YaqiWang) [inline framework/include/base/NonlinearSystem.h]: suggest to keep the previous line.

File: framework/include/base/ComputeDiracThread.h [hunk 3]
```diff
@@ -23,7 +23,7 @@
 #include "libmesh/elem_range.h"
 
 // Forward declarations
-class NonlinearSystem;
```
Comment (YaqiWang) [inline framework/include/base/ComputeDiracThread.h]: `MooseEigenSystem` is not changed in this commit?

**Comment by fdkong** [comment]
@friedmud @permcody, do you have any comments on this change?

**Comment by permcody** [comment]
I probably won't have time to review this PR until next week. Derek and @andrsd may want to look at this before it goes in. This is a big change.

**Comment by YaqiWang** [comment]
This is just step 1 and 2 and basically reshuffling the code.

**Comment by fdkong** [comment]
> I probably won't have time to review this PR until next week. Derek and @andrsd may want to look at this before it goes in. This is a big change.

This will be a big change in the future, but we are just reshuffling the code in this PR. This PR includes steps (1) and (2) only.

**Comment by friedmud** [comment]
I definitely want to look at it but I don't have time for it today (in the
middle of something else).

**Comment by friedmud** [comment]
I think I'm missing something.

Do you plan for `NonlinearSystem` to be _different_ from `NonlinearSystemBase` at some point?  Because in this PR... there is no purpose in splitting these (`NonlinearSystem` is just an empty shell).

If all you want to do is create a specialization of `NonlinearSystem` for your eigen solve... just inherit from `NonlinearSystem` and override virtuals.  No reason to create `NonlinearSystemBase`... but maybe I'm missing something...

**Comment by friedmud** [comment]
Ok - I see step (4) above... can you explain which methods you want to move from `NonlinearSystemBase` to `NonlinearSystem`?

Is there a reason they can't just be `virtual` in `NonlinearSystem` and then the eigen solver overrides them?

**Comment by fdkong** [comment]
> Ok - I see step (4) above... can you explain which methods you want to move from NonlinearSystemBase to NonlinearSystem?
> 
> Is there a reason they can't just be virtual in NonlinearSystem and then the eigen solver overrides them?

I do want to just inherit from NonlinearSystem, and then do whatever I want for eigenvalue problems. But there is one thing that stops me from doing this. The NonlinearSystem inherits from SystemTempl<TransientNonlinearImplicitSystem>, but I want MooseEigenSystem derive from SystemTempl<EigenSystem> so that I could use the eigen value interface in Libmesh. You possibly have better suggestions, @friedmud.

**Comment by fdkong** [comment]
@friedmud, there is another solution to get around this. We create an EigenSystem (LibMesh) Object inside the MooseEigenSystem, and then we override some functions from NonlinearSystem to use the eigen solvers. I not sure if it is a good idea.

**Comment by friedmud** [comment]
@fdkong will the eigenvalue solve need to solve _both_ a nonlinear system and an eigen system?

**Comment by fdkong** [comment]
> @fdkong will the eigenvalue solve need to solve both a nonlinear system and an eigen system?

@friedmud, good question. Usually not. The eigenvalue solve deals with the eigen system only. Sometimes, the eigenvalue system could be formed as a nonlinear system, and then use a normal nonlinear solver to find the eigen pair.

Ping @YaqiWang on this question.

**Comment by YaqiWang** [comment]
I think we hope not. We hope slepc can handle all different options of solving an eigenvalue problem, including using SNES to treat an eigenvalue problem as a nonlinear problem, doing some free power iterations to get a better initial guess for the SNES, handling other multiphysics variables, and more.

Currently eigen kernels do not have to live with eigen system. If they are on a nonlinear system, they behave like normal kernels except the scaling factor `1/lambda`, where lambda is given eigenvalue but not something we are going to seek.

It is not very clear to me on how this is going to be done in a right and better way.

**Comment by friedmud** [comment]
@fdkong : so is your idea to make `NonlinearSystemBase` a templated class so you can template it on the system type?

**Comment by fdkong** [comment]
> @fdkong : so is your idea to make NonlinearSystemBase a templated class so you can template it on the system type?

Yes, I hence could use the EigenSystem in Libmesh.

**Comment by fdkong** [comment]
> @fdkong : so is your idea to make NonlinearSystemBase a templated class so you can template it on the system type?
> 
> @friedmud. Yes, I hence could use the EigenSystem in Libmesh.

@friedmud, more suggestions on this?

**Comment by friedmud** [comment]
I still get the feeling that this isn't quite the right thing to do... but
maybe it will take some iteration.

No one else has anything to say?

**Comment by friedmud** [comment]
I feel like at the very least NonlinearSystemBase could be called something
else. AssemblySystem or something? Anyone have a better name?

The issue is that this inheritance tree is already pretty deep...

**Comment by fdkong** [comment]
> I feel like at the very least NonlinearSystemBase could be called something else. AssemblySystem or something? Anyone have a better name? The issue is that this inheritance tree is already pretty deep.

Agree, the inheritance is really deep, but no idea to remove this.

**Comment by permcody** [comment]
I agree with @friedmud's name. I don't like EigenSystem inheriting from
NonlinearSystemBase. Need a better name.

**Comment by fdkong** [comment]
> I feel like at the very least NonlinearSystemBase could be called something else. AssemblySystem or something? Anyone have a better name?

How about AssemblyEquations?

**Comment by andrsd** [comment]
A few comments:

1) IIRC to solve generalized eigen value problem you have to form 2 matrices. Back in Reno, when I was messing with SLEPc, that was the case. So, there has to be a syntax in the input file that will allow that, something that will tell FEProblem where to put each kernel.

2) the inheritance diagram of systems is as follows:
```
SystemBase -> SystemTemp<> +--> AuxiliarySystem
                           |
                           +--> NonlinearSystem
```
If the plan is to use EigenSystem from libmesh, we will have a 3rd branch like so:
```
SystemBase -> SystemTemp<> +--> AuxiliarySystem
                           |
                           +--> NonlinearSystem
                           |
                           +--> MooseEigenSystem
```
I do not see how `NonlinearSystemBase` will help here.

3) The major problem is that `FEProblem` is composed of aux and nl system. It almost feel like we need EigenProblem that is composed from EigenSystem and aux system. So I am thinking of creating something like:
```
TBDProblem +--> FEProblem
           |
           +--> EigenProblem
```

Move most of the current FEProblem functionality to `TBDProblem`, override the bits in their child classes to do the right solves, etc.

I do not believe that messing with the system hierarchy will lead to a good solution.

**Comment by permcody** [comment]
The intermediate base class was there to factor out commonality between `NonlinearSystem` and `MooseEigenSystem`. I believe there's a lot there that may not necessarily be in common with `AuxilarySystem`. I agree with the rest of your comments.

SystemBase -> SystemTemp<> +--> AuxiliarySystem
                           |
                           IntermediateBase+--> NonlinearSystem
                           |
                           +--> MooseEigenSystem

**Comment by permcody** [comment]
Forgot the fixed format tags. Diagram above should be fixed now.

**Comment by andrsd** [comment]
So, that is for the inverse power method, not SLEPc? Because IIRC, SLEPc needs 2 matrices, no vectors, so it so not similar to NonlinearSystem at all...

**Comment by permcody** [comment]
@fdkong thought there was overlap - If there's not then you are right and this isn't the correct solution. @fdkong?

**Comment by andrsd** [comment]
After checking SLEPc [examples](http://acts.nersc.gov/slepc/), they do 2 things:

1. Ax=!x, you need 1 matrix and an initial vector,
2. Ax=!Bx, you need 2 matrices.

I am assuming neutronics guys want (2).

The code path in both examples is:

1. form a matrix A, and B (if doing case 2)
2. call solve
3. get the eigen values back

Not sure how our solve options would apply here, meaning if PJFNK, JFNK has any meaning in this context...

**Comment by YaqiWang** [comment]
Our task in neutronics is to solve a generalized eigenvalue problem `Ax = 1/lambda Bx`, where `Bx` corresponds to eigen kernels, `lambda` is the eigenvalue. Typically we only cares the largest lambda and its eigenvector, i.e. the fundamental mode.

There are several things can make this task more complicated:
1. We typically cannot afford assembling the full `A` and `B` matrix. Thus we have to perform matrix-free calculations.
2. We currently have two eigen solvers, one is the inverse power method (iteratively keep solving `Ax = 1/lambda^previous Bx^previous`) another is PJFNK (using PETSc SNES). PJFNK requires a good initial guess which is typically obtained with few inverse power iterations. These two method proofed to be good options for most of neutronics applications. So we want the new eigen solver to preserve these two methods.
3. The eigen system could coupling with another physics, whose solution is not part of the eigen solution, i.e. arbitrarily scaling of its solution is not the solution of the system not like `x`.
4. Both `A` and `B` could have `x` dependency, i.e. the problem is nonlinear.
5. We could also solve the non-homogeneous eigen problem `Ax = 1/lambda Bx + b`, where `lambda` is the fundamental eigenvalue, making `A - 1/lambda B` a singular system. Apparently `b` must satisfy a condition `x^* dot b = 0`, here `x^*` is the adjoint fundamental eigen vector. And the number of solutions will be infinite, we have to have a condition to make the solution unique.
6. Similarly as 5, we could solve a transient system `dx/dt + Ax = 1/lambda Bx`, and the initial condition is set with the eigen solution.
7. Eigen system could be solved along a transient system, e.g. `dN/dt = alpha(x) N` and `A(N) x = 1/lambda B(N) x`. This is the case of depletion.
8. We sometimes care a value of parameter `p`, which affecting `A(p)` and `B(p)`. We want to adjust this value to make the eigenvalue `lambda` reach a certain value, typically 1 for critical. This kind of evaluations is known as criticality search.

I do not have much suggestions on the inheritance. I just noticed that `SystemTempl` is a template class. It can wrap any libmesh system including `EigenSystem`. I suggest we take this opportunity to clean up the structure of `NonlinearSystem` and `FEProblem` and shuffle the code to their right place.

**Comment by fdkong** [comment]
> A few comments:

Excellent comments. Thanks, @andrsd.

> 1) IIRC to solve generalized eigen value problem you have to form 2 matrices. Back in Reno, when I was messing with SLEPc, that was the case.

Yes, you are right. We need two matrices for generalized eigenvalue problems. There are two ways to implement: shellmatrix (matrix-free) and normal sparse matrix. 

> So, there has to be a syntax in the input file that will allow that, something that will tell FEProblem where to put each kernel.

We could do similar things like what we are doing for time kernel.

MooseSharedPointer\<EigenKernel\> eigen_kernel = MooseSharedNamespace::dynamic_pointer_cast\<EigenKernel\>(kernel);
    if (eigen_kernel)
      _right_eigen_kernel.addObject(kernel, tid);
    else
      _left_eigen_kernels.addObject(kernel, tid);

Kernels are stored separately in warehouses.  Which warehouse is used depends on which matrix we are working on.

> 
> 2) the inheritance diagram of systems is as follows:
> 
> SystemBase -> SystemTemp<> +--> AuxiliarySystem
>                            |
>                            +--> NonlinearSystem
> 
> If the plan is to use EigenSystem from libmesh, we will have a 3rd branch like so:
> 
> SystemBase -> SystemTemp<> +--> AuxiliarySystem
>                            |
>                            +--> NonlinearSystem
>                            |
>                            +--> MooseEigenSystem
> 
> I do not see how NonlinearSystemBase will help here.

The reason to create an intermediateSystem  because I want to reuse something from nonlinearSystem to avoid copying the code. For example, I could totally reuse computeResidualInternal when using shellMatrix, and reuse computeJacobianInternal when using a normal matrix or forming a preconditioner for eigen solvers. There are many things related to computing residual and Jaocibans, and I possibly could reuse most of them. 

The only thing that stops me from inheriting from the nonlinearSystem is that I want to bring  EigenSystem  from Libmesh  using SystemTemp\<EigenSystem\>. 

> 
> 3) The major problem is that FEProblem is composed of aux and nl system. It almost feel like we need EigenProblem that is composed from EigenSystem and aux system. So I am thinking of creating something like:
> 
> TBDProblem +--> FEProblem
>            |
>            +--> EigenProblem
> 

Good idea.  But the problem is that I can not distinguish the FEMProblem from NonlinearSystem much. Something, I think, should be moved to NonlinearSystem from FEMProblem. For example, FEProblem::checkNonlinearConvergence should live in  NonlinearSystem. 


> Move most of the current FEProblem functionality to TBDProblem, override the bits in their child classes to do the right solves, etc.

Yes, this will help.  Seperate Eigen from FEM. 

> 
> I do not believe that messing with the system hierarchy will lead to a good solution.

**Comment by fdkong** [comment]
> @fdkong thought there was overlap - If there's not then you are right and this isn't the correct solution. @fdkong?

There is some overlap, I believe

**Comment by fdkong** [comment]
> So, that is for the inverse power method, not SLEPc? Because IIRC, SLEPc needs 2 matrices, no vectors, so it is not similar to NonlinearSystem at all...

There are some vectors and callbacks  if we are using shellmatrix. @YaqiWang  likes to use the matrix-free.  It is not easy to form the exact Jacobian, sometime. 

One more thing, the matrix may be a nonlinear operator which depends on current solution (eigenvector).

**Comment by fdkong** [comment]
> After checking SLEPc examples, they do 2 things:
> 
>     Ax=!x, you need 1 matrix and an initial vector,
>     Ax=!Bx, you need 2 matrices.
> 
> I am assuming neutronics guys want (2).
> 
> The code path in both examples is:
> 
>     form a matrix A, and B (if doing case 2)
>     call solve
>     get the eigen values back
> 
> Not sure how our solve options would apply here, meaning if PJFNK, JFNK has any meaning in this context...

A lot of cases @YaqiWang  has are not "pure" eigen problems. There is an eigenvalue problem coupled with a normal equation, and it ends up to use JFNK solve the coupled nonlinear system with an initial guess obtained from eigensolvers. 

I am going to propose JFNK as an eigensolver in SLEPc, and we hence do not need to deal with this case in MOOSE.

**Comment by YaqiWang** [comment]
One more thing, #5288 appears related with this discussion.

**Comment by fdkong** [comment]
I like somethings:

```
SystemBase -> SystemTemp<> +--> AuxiliarySystem
                           |
                           +-->IntermediateBase+--> NonlinearSystem
                                               |
                                               +--> MooseEigenSystem
```

and 

```
TBDProblem +--> FEProblem
           |
           +--> EigenProblem
```

Any suggestions, @mooseframework?

**Comment by friedmud** [comment]
> Something, I think, should be moved to NonlinearSystem from FEMProblem. For example, FEProblem::checkNonlinearConvergence should live in NonlinearSystem.

I can tell you why that is the way it is: it's because a `Problem` can be chosen from the input file (thus allowing choosing a different convergence criteria from the input file) whereas a `System` cannot be chosen from the input file.

In my mind... the way to fix _that_ particular problem is to actually create a `ConvergenceChecker` type of object... that users can inherit from and override virtuals.  Then the `ConvergenceChecker` can be chosen from the input file (you could actually choose multiple if you wanted... so you could compose them).

That's a little off-topic... but I just wanted to point out that there _are_ some reasons why things are the way they are.

As for this particular issue: I think I'm ok with creating an intermediate base-class.  I'm also starting to believe that it _could_ be called `NonlinearSystemBase` too.

What is a `NonlinearSystem`?  To me... it's a system of nonlinear equations.  You need to be able to compute residuals and Jacobians to solve it.  That's exactly what this base class will be.  Now, _how_ you use those residuals/Jacobians is up to you... you may do some funky eigenvalue solve... or maybe you just use Newton's method.  But at its core you're still solving a (possibly) nonlinear system of equations.

So: My current vote is to merge the current pull request.

What say you @mooseframework ?

**Comment by friedmud** [comment]
Note: I'm probably ok with `AssemblySystem` as well.  It's a system that needs to "assemble" residuals and jacobians.

**Comment by fdkong** [comment]
> I can tell you why that is the way it is: it's because a Problem can be chosen from the input file (thus allowing choosing a different convergence criteria from the input file) whereas a System cannot be chosen from the input file.
> 
> In my mind... the way to fix that particular problem is to actually create a ConvergenceChecker type of object... that users can inherit from and override virtuals. Then the ConvergenceChecker can be chosen from the input file (you could actually choose multiple if you wanted... so you could compose them).
> 
> That's a little off-topic... but I just wanted to point out that there are some reasons why things are the way they are.
> 
> As for this particular issue: I think I'm ok with creating an intermediate base-class. I'm also starting to believe that it could be called NonlinearSystemBase too.
> 
> What is a NonlinearSystem? To me... it's a system of nonlinear equations. You need to be able to compute residuals and Jacobians to solve it. That's exactly what this base class will be. Now, how you use those residuals/Jacobians is up to you... you may do some funky eigenvalue solve... or maybe you just use Newton's method. But at its core you're still solving a (possibly) nonlinear system of equations.

@andrsd  also proposes another solution like this:

```
SystemBase -> SystemTemp<> +--> AuxiliarySystem
                           |
                           +--> NonlinearSystem  -- uses ----------- ComputeJacobianClass (Like a utility class)
                           |                                     |
                           +--> MooseEigenSystem ----also uses -----
```

ComputeJacobianClass (possibly another name) may include most of functions that MooseEigenSystem and NonlinearSystem can share. 

> So: My current vote is to merge the current pull request.
> 
> What say you @mooseframework ?

**Comment by friedmud** [comment]
I'm ok with that idea

**Comment by fdkong** [comment]
> I'm ok with that idea

@friedmud, Which one you prefer? The one proposed by @andrsd, possibly, is easier to implement. I am going to do some experiments.

**Comment by friedmud** [comment]
Let's see what @andrsd 's proposal would look like.

It would have the side benefit of breaking up NonlinearSystem (which is too
big anyway).

**Comment by fdkong** [comment]
> Let's see what @andrsd 's proposal would look like. It would have the side benefit of breaking up NonlinearSystem (which is too big anyway).

I am going to create another pull request next week or latter somehow.

**Comment by fdkong** [comment]
@andrsd, following your suggestions, I have a hierarchy:

```
SystemBase -> SystemTemp<> +--> AuxiliarySystem
                           |
                           +--> NonlinearSystem  <---+ AssemblySystem 
                           |                         |                
                           +--> MooseEigenSystem <----
```

In `AssemblySystem`, there is a function called `computeJacobianInternal` which, in turn, calls `ComputeJacobianThread cj(_fe_problem, *this, jacobian)` inheriting from ThreadedElementLoop.  Unfortunately, `ThreadedElementLoop` asks for an object of `SystemBase` as an input parameter in the construction function. We possibly could not simply change the class type,  `SystemBase`, to  `AssemblySystem` because some functions in `AuxiliarySystem` also use the class `ThreadedElementLoop`.

@andrsd, any suggestions on this?

**Comment by andrsd** [comment]
It looks like `_system` from `ThreadedElementLoop` is used only in `ComputeElemAuxVarsThread` and in that case `_system` is actually equal to `_aux_sys`. So it seems that you can get rid of `ThreadedElementLoop::_system` at all.

**Pull request closed without merging**